src/main/java/com/gw2auth/oauth2/server/service/gw2account/verification/Gw2AccountVerificationServiceImpl.java

```diff
@@ -243,23 +248,19 @@ public void cancelPendingChallenge(UUID accountId, UUID gw2AccountId) {
         }
     }
 
-    @Transactional(noRollbackFor = Gw2AccountVerificationServiceException.class)
-    public boolean verify(AccountService.LoggingContext logging, Gw2AccountVerificationChallengePendingEntity entity) {
+    private boolean verifyInternal(AccountService.LoggingContext logging,
+                                   Gw2AccountVerificationChallengePendingEntity entity,
+                                   VerificationChallenge<?> challenge) throws IOException {
+
         final UUID accountId = entity.accountId();
         final UUID gw2AccountId = entity.gw2AccountId();
 
-        final VerificationChallenge<?> challenge = this.challengesById.get(entity.challengeId());
-        if (challenge == null) {
-            this.gw2AccountVerificationChallengePendingRepository.deleteByAccountIdAndGw2AccountId(accountId, gw2AccountId);
-            throw new Gw2AccountVerificationServiceException(Gw2AccountVerificationServiceException.CHALLENGE_NOT_FOUND, HttpStatus.NOT_FOUND);
-        }
-
         final boolean isVerified;
         try {
-            isVerified = verify(challenge, entity.state(), entity.gw2ApiToken());
+            isVerified = readStateAndVerify(challenge, entity.state(), entity.gw2ApiToken());
```

Comment:
Since now `verifyInternal` receives the `VerificationChallenge` object, this method can be made generic instead of `readStateAndVerify` and we could reduce the scope of this try-catch to only `challenge.readState` instead of covering both `readState` and `verify`

Reply:
Solved this by creating an overload of the `verify` method that takes a string. This way, we can avoid the "hacky" generics.